web/src/main/java/com/microsoft/applicationinsights/web/internal/httputils/HttpServerHandler.java

```diff
@@ -0,0 +1,131 @@
+package com.microsoft.applicationinsights.web.internal.httputils;
+
+import com.microsoft.applicationinsights.TelemetryClient;
+import com.microsoft.applicationinsights.common.CommonUtils;
+import com.microsoft.applicationinsights.internal.logger.InternalLogger;
+import com.microsoft.applicationinsights.telemetry.Duration;
+import com.microsoft.applicationinsights.telemetry.RequestTelemetry;
+import com.microsoft.applicationinsights.web.internal.RequestTelemetryContext;
+import com.microsoft.applicationinsights.web.internal.ThreadContext;
+import com.microsoft.applicationinsights.web.internal.WebModulesContainer;
+import org.apache.commons.lang3.Validate;
+import org.apache.commons.lang3.exception.ExceptionUtils;
+import org.apache.http.annotation.Experimental;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.net.MalformedURLException;
+import java.util.Date;
+
+/**
+ * This Helper Handler class provides the required methods to instrument requests.
+ * @param <P> The HttpRequest entity
+ * @param <Q> The HttpResponse entity
+ */
+@Experimental
+public final class HttpServerHandler<P /* >>> extends @NonNull Object */, Q> {
```

Comment:
Should we provide an interface for this class to implement which contains onEnd and onBegin methods. Would it make sense to have generic Handler interface and this class implement it? However the question would be how many arguments would be passed to the generic interface methods? For http we need 2, but might need more or less for others? One approach can be to use varargs and fetch parameters from there but then methods needs to have have hard logic inside them for ordering of parameters inside varargs array.